### DISCUSSION TITLE
DrawerControl

### DISCUSSION THREAD

**Discussion by Xiaoy312** [unoplatform/uno.toolkit.ui#51] [Ideas]
> This control is a control that allows you to use swipe gesture opens another secondary view on top of primary view.
> It is designed to be integrated into SplitView and by extension into NavigationView to enable gesture support. It can also be used as it is.

![20210723 drawer control](https://user-images.githubusercontent.com/2359550/126802535-e0d25abb-5ffa-42e6-9a28-a592bff04c87.gif)

The contract for the control is described below:
Type|Property Name|Description
-|-|-
object|DrawerContent|Content to display in the drawer (secondary view)
double|DrawerDepth|Height or width (depending on the OpenDirection) of the drawer
DrawerOpenDirection<br>_enum: {Right,Left,Down,Up}_|OpenDirection|Direction in which the drawer opens.
bool|IsOpen|self-explanatory
Brush|LightDismissOverlayBackground|Brush used to paint the light dismiss area. It will be animated between 0~100% opacity as the drawer closes/opens.
double?|EdgeSwipeDetectionLength|Defines the area (horizontal: width, vertical: height) that can be used to initiate the opening swipe gesture. `null` for from anywhere.
bool=true|IsGestureEnabled|Indicates if the gesture events will be handled

> There is an extension class `DrawerControlBehavior` for adjusting some of above properties through attached properties when nested in another control's template. This is done because certain properties don't fit well into `SplitView` and `NavigationView` contract, like: OpenDirection, LightDismissOverlayBackground, EdgeSwipeDetectionLength, IsGestureEnabled.

Constants:
Type|Property Name|Description
-|-|-
double= 1.0 / 3|DragToggleThresholdRatio|Distance (based on `DrawerDepth`) needed to swipe open/close the drawer
double= 0.95|AnimateSnappingThresholdRatio|If the drawer is already very close (95% way there) to the snap point, then no animation is played...
TimeSpan= 150ms|AnimationDuration|...else, a flat 150ms animation is played to finish up the work

```

**Comment by Xiaoy312** [comment]
There are currently some limitations/issues to the control itself:
- If contained inside a ScrollViewer, the DrawerOpenDirection cannot be parallel to the ScrollViewer (but usually not an issue, as this should be a top level control)
- ios: https://github.com/unoplatform/uno/issues/5385

**Comment by Xiaoy312** [comment]
Should we split `EdgeSwipeDetectionLength` (`double?`) into 
`UseEdgeSwipeToOpen` (`bool`) & `EdgeSwipeDetectionLength` (`double`)?

**Reply by davidjohnoliver** [reply]
You can just use `double.NaN` to indicate that it's unset, it doesn't need to be nullable.
